### PR TITLE
transport/common: Share DNS lookups between TCP and WebSocket

### PR DESCRIPTION
**Pull request by lexnv**
Move the dns resolving to a dedicated common module and return specific errors on failures. 

Part of: https://github.com/paritytech/litep2p/issues/70

### CODE REVIEW

File: src/transport/common/listener.rs [hunk 1]
```diff
@@ -311,10 +367,27 @@ fn multiaddr_to_socket_address(
...
         },
-        Some(Protocol::Dns(address))
-        | Some(Protocol::Dns4(address))
-        | Some(Protocol::Dns6(address)) => match iter.next() {
-            Some(Protocol::Tcp(port)) => AddressType::Dns(address.to_string(), port),
+        Some(Protocol::Dns(address)) | Some(Protocol::Dns4(address)) => match iter.next() {
+            Some(Protocol::Tcp(port)) => AddressType::Dns {
+                address: address.into(),
+                port,
+                is_dns6: false,
```
Comment (dmitry-markin) [inline src/transport/common/listener.rs:374]: As [multiaddr spec](`https://github.com/multiformats/multiaddr/blob/master/protocols.csv`) says, `Protocol::Dns` can resolve to both IPv4 & IPv6. I would use this PR as an opportunity to fix the existing behavior and allow IPv6 addresses with `Protocol::Dns`.

**Review by dmitry-markin — APPROVED** [review]
LGTM! :+1:

File: src/transport/common/listener.rs (hunk 1 above)
Comment (lexnv) [reply src/transport/common/listener.rs:374]: Nice catch! For sure, will fix it right away

File: src/transport/tcp/connection.rs [hunk 2]
```diff
@@ -455,9 +459,21 @@ impl TcpConnection {
...
+                DnsType::Dns => Multiaddr::empty()
+                    .with(Protocol::Dns(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+                DnsType::Dns4 => Multiaddr::empty()
+                    .with(Protocol::Dns4(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+                DnsType::Dns6 => Multiaddr::empty()
+                    .with(Protocol::Dns6(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+            },
```
Comment (lexnv) [inline src/transport/tcp/connection.rs:476]: Here we had another miss-match, where regardless of the initial DNS version we'd accept both ipv4 or ipv6

**Review by dmitry-markin — APPROVED** [review]
Nice!

**Pull request merged**